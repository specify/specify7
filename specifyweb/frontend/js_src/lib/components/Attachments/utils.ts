import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
