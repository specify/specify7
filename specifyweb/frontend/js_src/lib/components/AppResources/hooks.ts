import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { useAsyncState } from '../../hooks/useAsyncState';
import { f } from '../../utils/functools';
import type { GetOrSet, IR, RA } from '../../utils/types';
import { addMissingFields } from '../DataModel/addMissingFields';
import { fetchCollection } from '../DataModel/collection';
import type { SerializedResource } from '../DataModel/helperTypes';
import type {
  Collection,
  Discipline,
  SpAppResource,
  SpAppResourceData,
  SpecifyUser,
  SpViewSetObj,
} from '../DataModel/types';
import { userPreferences } from '../Preferences/userPreferences';
import { getAppResourceType } from './filtersHelpers';
import { getAppResourceCount, getAppResourceMode } from './helpers';
import { getAppResourceTree, getScope } from './tree';
import type { ScopedAppResourceDir } from './types';
import { appResourceSubTypes } from './types';

export type AppResources = {
  readonly directories: RA<ScopedAppResourceDir>;
  readonly disciplines: RA<SerializedResource<Discipline>>;
  readonly collections: RA<SerializedResource<Collection>>;
  readonly users: RA<SerializedResource<SpecifyUser>>;
  readonly appResources: RA<SerializedResource<SpAppResource>>;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
};

export function useAppResources(
  loadingScreen: boolean = true
): GetOrSet<AppResources | undefined> {
  return useAsyncState(
    React.useCallback(
      async () =>
        f.all({
          directories: fetchCollection('SpAppResourceDir', {
            limit: 0,
            domainFilter: false,
          }).then<AppResources['directories']>(({ records }) =>
            records.map((record) => ({ ...record, scope: getScope(record) }))
          ),
          disciplines: fetchCollection('Discipline', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          collections: fetchCollection('Collection', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          users: fetchCollection('SpecifyUser', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          appResources: fetchCollection('SpAppResource', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          viewSets: fetchCollection('SpViewSetObj', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
        }),
      []
    ),
    loadingScreen
  );
}

export type AppResourcesTree = RA<{
  readonly label: LocalizedString;
  /*
   * A string that would be stable thought the lifecycle of an object.
   * Used to identify a tree node when storing conformation it in localStorage.
   *
   * Directory key is created instead of using id, because app resource
   * needs a way to identify directories that don't yet have IDs (i.e, when
   * creating an app resource for a collection that didn't have any app resources
   * before, the collection won't have any SpAppResourceDir record, thus not
   * directory ID).
   */
  readonly key: string;
  readonly directory: ScopedAppResourceDir | undefined;
  readonly appResources: RA<
    SerializedResource<SpAppResource> & {
      readonly label?: LocalizedString;
    }
  >;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
  readonly subCategories: AppResourcesTree;
}>;

export function useResourcesTree(resources: AppResources): AppResourcesTree {
  const [localize] = userPreferences.use(
    'appResources',
    'appearance',
    'localizeResourceNames'
  );
  return React.useMemo<AppResourcesTree>(() => {
    const tree = getAppResourceTree(resources);
    return localize ? localizeTree(tree) : tree;
  }, [resources, localize]);
}

const localizeTree = (tree: AppResourcesTree): AppResourcesTree =>
  tree.map(({ appResources, subCategories, ...rest }) => ({
    ...rest,
    appResources: appResources.map(localizeResource),
    subCategories: localizeTree(subCategories),
  }));

function localizeResource(
  resource: SerializedResource<SpAppResource> & {
    readonly label?: LocalizedString;
  }
): SerializedResource<SpAppResource> & { readonly label?: LocalizedString } {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  // Check that resource of this type can only have one specific name
  return typeof type.name === 'string'
    ? {
        ...resource,
        // Then replace the name with a localized label unless it's already set
        label: resource.label ?? type.label,
      }
    : resource;
}

export function useAppResourceCount(
  resourcesTree: AppResourcesTree[number]
): number {
  return React.useMemo(
    () => getAppResourceCount(resourcesTree),
    [resourcesTree]
  );
}

/**
 * Fetch resource contents
 */
export function useAppResourceData(
  resource: SerializedResource<SpAppResource | SpViewSetObj>,
  initialData: string | undefined
): {
  readonly resourceData: GetOrSet<
    SerializedResource<SpAppResourceData> | undefined
  >;
  readonly isChanged: boolean;
} {
  const initialValue = React.useRef<string>('');
  const [resourceData, setResourceData] = useAsyncState<
    SerializedResource<SpAppResourceData>
  >(
    React.useCallback(async () => {
      const relationshipName =
        getAppResourceMode(resource) === 'appResources'
          ? 'spAppResource'
          : 'spViewSetObj';
      const newResource = addMissingFields('SpAppResourceData', {
        [relationshipName]: resource.id,
        data: initialData,
      });
      const dataResource =
        typeof resource.id === 'number'
          ? await fetchCollection('SpAppResourceData', {
              limit: 1,
              [relationshipName]: resource.id,
              domainFilter: false,
            }).then(
              ({ records }) =>
                /*
                 * For some reason, app resource can have multiple app resource
                 * datas (but it never does in practice)
                 */
                records[0] ?? newResource
            )
          : newResource;
      const newData = fixLineBreaks(dataResource.data ?? '');
      initialValue.current = newData;
      return { ...dataResource, data: newData };
    }, [resource, initialData]),
    true
  );
  return {
    resourceData: [resourceData, setResourceData],
    isChanged: initialValue.current !== resourceData?.data,
  };
}

const fixLineBreaks = (string: string): string =>
  string.replaceAll(/[\n\r]+/gu, '\n');

export const getAppResourceExtension = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): string =>
  resource._tableName === 'SpViewSetObj'
    ? 'xml'
    : getResourceExtension(resource as SerializedResource<SpAppResource>);

function getResourceExtension(
  resource: SerializedResource<SpAppResource>
): 'jrxml' | 'json' | 'properties' | 'txt' | 'xml' {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  const mimeType = resource.mimeType?.toLowerCase() ?? type?.mimeType ?? '';
  if (mimeType in mimeMapper) return mimeMapper[mimeType];
  else if (mimeType.startsWith('jrxml')) return 'jrxml';
  else if (
    resource.name === 'preferences' &&
    mimeType === appResourceSubTypes.otherPropertiesResource.mimeType
  )
    return 'properties';
  else return 'txt';
}

const mimeMapper: IR<'json' | 'properties' | 'txt' | 'xml'> = {
  'application/json': 'json',
  'text/xml': 'xml',
  'text/plain': 'txt',
};
