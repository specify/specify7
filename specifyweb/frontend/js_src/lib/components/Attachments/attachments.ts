import { commonText } from '../../localization/common';
import { ajax } from '../../utils/ajax';
import { Http } from '../../utils/ajax/definitions';
import { handleAjaxResponse } from '../../utils/ajax/response';
import type { IR, RA } from '../../utils/types';
import type { UploadAttachmentSpec } from '../AttachmentsBulkImport/types';
import { getField } from '../DataModel/helpers';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { tables } from '../DataModel/tables';
import type { Attachment } from '../DataModel/types';
import { load } from '../InitialContext';
import { getIcon, unknownIcon } from '../InitialContext/icons';
import { getPref } from '../InitialContext/remotePrefs';
import { formatUrl } from '../Router/queryString';

type AttachmentSettings = {
  readonly collection: string;
  readonly delete: string;
  readonly getmetadata: string;
  readonly read: string;
  readonly testkey: string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  readonly token_required_for_get: boolean;
  readonly write: string;
};

let settings: AttachmentSettings | undefined;
export const attachmentSettingsPromise = load<AttachmentSettings | IR<never>>(
  '/context/attachment_settings.json',
  'application/json'
).then((data) => {
  if (Object.keys(data).length > 0) settings = data as AttachmentSettings;
  return attachmentsAvailable();
});

export const attachmentsAvailable = (): boolean => typeof settings === 'object';

const thumbnailable = new Set([
  'image/jpeg',
  'image/gif',
  'image/png',
  'image/tiff',
  'application/pdf',
]);

function iconForMimeType(mimeType: string): {
  readonly alt: string;
  readonly src: string;
} {
  if (mimeType === 'text/plain')
    return { alt: 'text', src: getIcon('text') ?? unknownIcon };
  if (mimeType === 'text/html')
    return { alt: 'html', src: getIcon('html') ?? unknownIcon };

  const parts = mimeType.split('/');
  const type = parts[0];
  const subtype = parts[1];

  if (['audio', 'video', 'image', 'text'].includes(type))
    return { alt: type, src: getIcon(type) ?? unknownIcon };

  if (type === 'application') {
    const iconName = {
      pdf: 'pdf',
      'vnd.ms-excel': 'MSExcel',
      'vnd.ms-word': 'MSWord',
      'vnd.ms-powerpoint': 'MSPowerPoint',
    }[subtype];

    if (typeof iconName === 'string')
      return { alt: iconName, src: getIcon(iconName) ?? unknownIcon };
  }

  return { alt: commonText.unknown(), src: getIcon('unknown') ?? unknownIcon };
}

export const fetchAssetToken = async (
  fileName: string,
  silent: boolean = false
) =>
  ajax(formatUrl('/attachment_gw/get_token/', { fileName }), {
    headers: { Accept: 'text/plain' },
    errorMode: silent ? 'silent' : 'dismissible',
    expectedErrors: silent ? Object.values(Http) : [Http.OK],
  }).then(({ data, status }) => (status === Http.OK ? data : undefined));

const fetchToken = async (fileName: string): Promise<string | undefined> =>
  settings?.token_required_for_get === true
    ? fetchAssetToken(fileName)
    : Promise.resolve(undefined);

export type AttachmentThumbnail = {
  readonly src: string;
  readonly alt: string | undefined;
  readonly width: number;
  readonly height: number;
};

export async function fetchThumbnail(
  attachment: SerializedResource<Attachment>,
  scale = getPref('attachment.preview_size')
): Promise<AttachmentThumbnail | undefined> {
  const mimeType = attachment.mimeType ?? undefined;
  const thumbnail =
    mimeType === undefined || thumbnailable.has(mimeType)
      ? undefined
      : iconForMimeType(mimeType);

  // Display an icon for resources that don't have a custom thumbnail
  if (typeof thumbnail === 'object' && thumbnail?.src !== unknownIcon)
    return {
      ...thumbnail,
      width: scale,
      height: scale,
    };

  // Fetch a preview for resources that support thumbnail
  if (attachment.attachmentLocation === null || settings === undefined)
    return undefined;
  const token = await fetchToken(attachment.attachmentLocation);
  return {
    src: formatUrl(settings.read, {
      coll: settings.collection,
      type: 'T',
      fileName: attachment.attachmentLocation,
      scale,
      token,
    }),
    alt: attachment.attachmentLocation ?? undefined,
    width: scale,
    height: scale,
  };
}

export const formatAttachmentUrl = (
  attachment: SerializedResource<Attachment>,
  token: string | undefined
): string | undefined =>
  typeof settings === 'object'
    ? formatUrl(settings.read, {
        coll: settings.collection,
        type: 'O',
        fileName: attachment.attachmentLocation,
        downloadName: attachment.origFilename?.replace(/^.*[/\\]/u, ''),
        token,
      })
    : undefined;

export const fetchOriginalUrl = async (
  attachment: SerializedResource<Attachment>
): Promise<string | undefined> =>
  typeof attachment.attachmentLocation === 'string'
    ? fetchToken(attachment.attachmentLocation).then((token) =>
        formatAttachmentUrl(attachment, token)
      )
    : Promise.resolve(undefined);

export async function uploadFile(
  file: File,
  handleProgress: (percentage: number | true) => void,
  uploadAttachmentSpec?: UploadAttachmentSpec,
  strict = true
): Promise<SpecifyResource<Attachment> | undefined> {
  if (settings === undefined) return undefined;

  const data =
    typeof uploadAttachmentSpec === 'object'
      ? uploadAttachmentSpec
      : await ajax<RA<Partial<UploadAttachmentSpec>>>(
          '/attachment_gw/get_upload_params/',
          {
            method: 'POST',
            headers: { Accept: 'application/json' },
            body: {
              filenames: [file.name],
            },
          }
        ).then(({ data }) => data[0]);

  if (data.attachmentLocation === undefined || data.token === undefined)
    return undefined;

  const formData = new FormData();

  formData.append('file', file);
  formData.append('token', data.token);
  formData.append('store', data.attachmentLocation);
  formData.append('type', 'O');
  formData.append('coll', settings.collection);

  /*
   * Using XMLHttpRequest rather than fetch() because need upload
   * progress reporting, which is not yet supported by fetch API
   */

  const xhr = new XMLHttpRequest();
  xhr.upload?.addEventListener('progress', (event) =>
    handleProgress(event.lengthComputable ? event.loaded / event.total : true)
  );
  xhr.open('POST', settings.write);
  xhr.send(formData);
  const DONE = 4;
  await new Promise((resolve, reject) =>
    xhr.addEventListener('readystatechange', () => {
      if (xhr.readyState === DONE)
        try {
          resolve(
            handleAjaxResponse({
              expectedErrors: [],
              accept: undefined,
              // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
              response: {
                ok: xhr.status === Http.OK,
                status: xhr.status,
                url: settings!.write,
              } as Response,

              errorMode: strict ? 'visible' : 'silent',
              text: xhr.responseText,
            })
          );
        } catch (error) {
          reject(error);
        }
    })
  );
  return new tables.Attachment.Resource({
    attachmentlocation: data.attachmentLocation,
    mimetype: fixMimeType(file.type),
    origfilename: file.name,
    title: file.name,
    isPublic: getPref('attachment.is_public_default'),
  });
}

/**
 * A temporary workaround for mimeTypes for `.docx` and `.xlsx` files being
 * longer than the length limit on the `Attachment.mimeType` field.
 * See: https://github.com/specify/specify7/issues/1141
 * REFACTOR: remove this once that issue is fixed
 */
function fixMimeType(originalMimeType: string): string {
  const maxLength = getField(tables.Attachment, 'mimeType').length;
  if (maxLength === undefined || originalMimeType.length < maxLength)
    return originalMimeType;
  else {
    const mimeType = 'application/octet-stream';
    console.warn(
      `Shortened the Attachment mimeType from "${originalMimeType}"` +
        ` to "${mimeType}" due to length limit`
    );
    return mimeType;
  }
}
