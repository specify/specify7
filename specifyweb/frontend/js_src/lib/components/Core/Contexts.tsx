import React from 'react';

import { useBooleanState } from '../../hooks/useBooleanState';
import { useCachedState } from '../../hooks/useCachedState';
import { commonText } from '../../localization/common';
import type { RA, WritableArray } from '../../utils/types';
import { setDevelopmentGlobal } from '../../utils/types';
import { error } from '../Errors/assert';
import { crash } from '../Errors/Crash';
import { ErrorBoundary } from '../Errors/ErrorBoundary';
import { Toasts } from '../Errors/Toasts';
import { loadingBar } from '../Molecules';
import { Dialog, dialogClassNames, LoadingScreen } from '../Molecules/Dialog';
import { TooltipManager } from '../Molecules/Tooltips';
import {
  SetUnloadProtectsContext,
  UnloadProtectsContext,
  UnloadProtectsRefContext,
} from '../Router/UnloadProtect';

// Stores errors that occurred before <Context> is rendered
const pendingErrors: WritableArray<ErrorComponent> = [];
type ErrorComponent = (props: { readonly onClose: () => void }) => JSX.Element;
let setError: (error: ErrorComponent) => void;

/**
 * Allows to display an error dialog from anywhere
 */
export function displayError(error: ErrorComponent): void {
  if (typeof setError === 'function') setError(error);
  else pendingErrors.push(error);
}

// This preserves the error messages even if <Context> gets re-rendered
let globalErrors: RA<JSX.Element> = [];

/*
 * For usage in non-react components only
 * REFACTOR: remove this once everything is using react
 */
let legacyContext: (promise: Promise<unknown>) => void;
// eslint-disable-next-line functional/prefer-tacit
export const legacyLoadingContext = (promise: Promise<unknown>): void =>
  legacyContext(promise);

/**
 * Provide contexts used by other components
 *
 * It is best practice to use context as little as possible, as they make
 * components more dependent on their parents.
 *
 * Thus, contexts were used only when necessary, and defined as higher up
 * the tree as possible, so that code refactoring does not lead to a
 * situation where context is accessed before it is defined.
 *
 * Defining contexts very high also allows the top ErrorBoundary to have
 * access to them.
 */
export function Contexts({
  children,
}: {
  readonly children: JSX.Element | RA<JSX.Element>;
}): JSX.Element {
  // Loading Context
  const [isLoading, handleLoading, handleLoaded] = useBooleanState();
  const loadingHandler = useLoadingLogic(handleLoading, handleLoaded);
  legacyContext = loadingHandler;

  // Error Context
  const [errors, setErrors] = React.useState<RA<JSX.Element>>(globalErrors);
  const handleError = React.useCallback(
    (error: (props: { readonly onClose: () => void }) => JSX.Element) =>
      setErrors((errors) => {
        const newError = (
          <React.Fragment key={errors.length}>
            {error({
              onClose: () =>
                setErrors((newErrors) =>
                  newErrors.filter((error) => error !== newError)
                ),
            })}
          </React.Fragment>
        );
        const newErrors = [...errors, newError];
        globalErrors = newErrors;
        return newErrors;
      }),
    []
  );
  if (setError === undefined) pendingErrors.forEach(handleError);
  setError = handleError;

  const [unloadProtects, setUnloadProtects] = React.useState<RA<string>>([]);

  const unloadProtectsRef = React.useRef(unloadProtects);
  const handleChangeUnloadProtects = React.useCallback(
    (value: RA<string> | ((oldValue: RA<string>) => RA<string>)): void => {
      const resolvedValue =
        typeof value === 'function' ? value(unloadProtectsRef.current) : value;
      setUnloadProtects(resolvedValue);
      unloadProtectsRef.current = resolvedValue;

      setDevelopmentGlobal('_unloadProtects', resolvedValue);
    },
    []
  );

  const isReadOnly = React.useContext(ReadOnlyContext);
  const isReadOnlyMode = useCachedState('forms', 'readOnlyMode')[0] ?? false;
  return (
    <UnloadProtectsContext.Provider value={unloadProtects}>
      <UnloadProtectsRefContext.Provider value={unloadProtectsRef}>
        <SetUnloadProtectsContext.Provider value={handleChangeUnloadProtects}>
          <Toasts>
            <ErrorBoundary>
              <ErrorContext.Provider value={handleError}>
                {errors}
                <LoadingContext.Provider value={loadingHandler}>
                  {isLoading && (
                    <Dialog
                      buttons={undefined}
                      className={{
                        container: dialogClassNames.narrowContainer,
                      }}
                      header={commonText.loading()}
                      onClose={undefined}
                    >
                      {loadingBar}
                    </Dialog>
                  )}
                  <React.Suspense fallback={<LoadingScreen />}>
                    <ReadOnlyContext.Provider
                      value={isReadOnly || isReadOnlyMode}
                    >
                      {children}
                    </ReadOnlyContext.Provider>
                  </React.Suspense>
                </LoadingContext.Provider>
                <TooltipManager />
              </ErrorContext.Provider>
            </ErrorBoundary>
          </Toasts>
        </SetUnloadProtectsContext.Provider>
      </UnloadProtectsRefContext.Provider>
    </UnloadProtectsContext.Provider>
  );
}

/**
 * Wait 50ms before displaying loading screen
 *   -> to avoid blinking a loading screen for resolved promises
 *      (that can also trigger bugs, like this one:
 *      https://github.com/specify/specify7/issues/884#issuecomment-1509324664)
 * Wait 50sm before removing loading screen
 *   -> to avoid flashing the screen when one loading screen is immediately
 *      followed by another one
 * 50ms was chosen as the longest delay that I don't notice in comparison to 0ms
 * (on an fast macbook pro with high refresh rate). The value might have to be
 * adjusted in the future
 */
const loadingScreenDelay = 50;

export function useLoadingLogic(
  handleLoading: () => void,
  handleLoaded: () => void
): (promise: Promise<unknown>) => void {
  const holders = React.useRef<RA<number>>([]);
  const loadingTimeout = React.useRef<
    ReturnType<typeof setTimeout> | undefined
  >(undefined);
  return React.useCallback(
    (promise: Promise<unknown>): void => {
      const holderId = Math.max(-1, ...holders.current) + 1;
      holders.current = [...holders.current, holderId];
      clearTimeout(loadingTimeout.current);
      loadingTimeout.current = setTimeout(handleLoading, loadingScreenDelay);
      promise
        .finally(() => {
          holders.current = holders.current.filter((item) => item !== holderId);
          if (holders.current.length > 0) return;
          clearTimeout(loadingTimeout.current);
          loadingTimeout.current = setTimeout(handleLoaded, loadingScreenDelay);
        })
        .catch(crash);
    },
    [handleLoading, handleLoaded]
  );
}

/*
 * REFACTOR: consider turning LoadingContext and useNavigate into global
 *   functions since they have the same value in all components and never change
 */
/**
 * Display a modal loading dialog while promise is resolving.
 * Also, catch and handle erros if promise is rejected.
 * If multiple promises are resolving at the same time, the dialog is
 * visible until all promises are resolved.
 * This prevents having more than one loading dialog visible at the same time.
 */
export const LoadingContext = React.createContext<
  (promise: Promise<unknown>) => void
>(() => error('Not defined'));
LoadingContext.displayName = 'LoadingContext';

/** Display a modal error message dialog */
export const ErrorContext = React.createContext<
  (error: (props: { readonly onClose: () => void }) => JSX.Element) => void
>(() => error('Not defined'));
ErrorContext.displayName = 'ErrorContext';

/** If true, renders everything below it as read only */
export const ReadOnlyContext = React.createContext<boolean>(false);
ReadOnlyContext.displayName = 'ReadOnlyContext';

/** If true, form is rendered in a search dialog - required fields are not enforced */
export const SearchDialogContext = React.createContext<boolean>(false);
SearchDialogContext.displayName = 'SearchDialogContext';
