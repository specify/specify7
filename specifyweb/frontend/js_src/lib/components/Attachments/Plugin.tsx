/**
 * Attachments form plugin
 */

import React from 'react';

import { useAsyncState, usePromise } from '../../hooks/useAsyncState';
import { useBooleanState } from '../../hooks/useBooleanState';
import { useErrorContext } from '../../hooks/useErrorContext';
import { useTriggerState } from '../../hooks/useTriggerState';
import { attachmentsText } from '../../localization/attachments';
import { formsText } from '../../localization/forms';
import { f } from '../../utils/functools';
import type { GetOrSet } from '../../utils/types';
import { Progress } from '../Atoms';
import { LoadingContext, ReadOnlyContext } from '../Core/Contexts';
import { toTable } from '../DataModel/helpers';
import type { AnySchema } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import type { Attachment } from '../DataModel/types';
import { raise } from '../Errors/Crash';
import { loadingBar } from '../Molecules';
import { Dialog } from '../Molecules/Dialog';
import { FilePicker } from '../Molecules/FilePicker';
import { ProtectedTable } from '../Permissions/PermissionDenied';
import { AttachmentPluginSkeleton } from '../SkeletonLoaders/AttachmentPlugin';
import { attachmentSettingsPromise, uploadFile } from './attachments';
import { AttachmentViewer } from './Viewer';

export function AttachmentsPlugin(
  props: Parameters<typeof ProtectedAttachmentsPlugin>[0]
): JSX.Element | null {
  const [available] = usePromise(attachmentSettingsPromise, true);
  return available === undefined ? null : available ? (
    <ProtectedTable action="read" tableName="Attachment">
      <ProtectedAttachmentsPlugin {...props} />
    </ProtectedTable>
  ) : (
    <p>{attachmentsText.attachmentServerUnavailable()}</p>
  );
}

/** Retrieve attachment related to a given resource */
export function useAttachment(
  resource: SpecifyResource<AnySchema> | undefined
): GetOrSet<SpecifyResource<Attachment> | false | undefined> {
  return useAsyncState(
    React.useCallback(
      async () =>
        f.maybe(resource, (resource) => toTable(resource, 'Attachment')) ??
        (await resource?.rgetPromise('attachment')) ??
        false,
      [resource]
    ),
    false
  );
}

function ProtectedAttachmentsPlugin({
  resource,
}: {
  readonly resource: SpecifyResource<AnySchema> | undefined;
}): JSX.Element | null {
  const [attachment, setAttachment] = useAttachment(resource);
  const isReadOnly = React.useContext(ReadOnlyContext);

  useErrorContext('attachment', attachment);

  const filePickerContainer = React.useRef<HTMLDivElement | null>(null);
  const related = useTriggerState(
    resource?.specifyTable.name === 'Attachment' ? undefined : resource
  );
  return attachment === undefined ? (
    <AttachmentPluginSkeleton />
  ) : (
    <div
      className="flex h-full gap-8 overflow-x-auto"
      ref={filePickerContainer}
      tabIndex={-1}
    >
      {typeof attachment === 'object' ? (
        <AttachmentViewer
          attachment={attachment}
          related={related}
          onViewRecord={undefined}
        />
      ) : isReadOnly ? (
        <p>{formsText.noData()}</p>
      ) : (
        <UploadAttachment
          onUploaded={(attachment): void => {
            // Fix focus loss when <FilePicker would be removed from DOM
            filePickerContainer.current?.focus();
            if (typeof resource === 'object')
              attachment?.set('tableID', resource.specifyTable.tableId);
            resource?.set('attachment', attachment as never);
            setAttachment(attachment);
          }}
        />
      )}
    </div>
  );
}

export function UploadAttachment({
  onUploaded: handleUploaded,
}: {
  readonly onUploaded: (attachment: SpecifyResource<Attachment>) => void;
}): JSX.Element {
  const [uploadProgress, setUploadProgress] = React.useState<
    number | true | undefined
  >(undefined);
  const [isFailed, handleFailed] = useBooleanState();
  const loading = React.useContext(LoadingContext);

  return isFailed ? (
    <p>{attachmentsText.attachmentServerUnavailable()}</p>
  ) : typeof uploadProgress === 'object' ? (
    <Dialog
      buttons={undefined}
      header={attachmentsText.uploadingInline()}
      onClose={undefined}
    >
      <div aria-live="polite">
        {typeof uploadProgress === 'number' ? (
          <Progress value={uploadProgress} />
        ) : (
          loadingBar
        )}
      </div>
    </Dialog>
  ) : (
    <FilePicker
      acceptedFormats={undefined}
      onFileSelected={(file): void =>
        loading(
          uploadFile(file, setUploadProgress)
            .then((attachment) =>
              attachment === undefined
                ? handleFailed()
                : handleUploaded(attachment)
            )
            .catch((error) => {
              handleFailed();
              raise(error);
            })
            .finally(() => setUploadProgress(undefined))
        )
      }
    />
  );
}
