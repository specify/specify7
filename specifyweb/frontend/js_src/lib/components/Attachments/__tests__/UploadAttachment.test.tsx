import { fireEvent, waitFor } from '@testing-library/react';
import React from 'react';

import { clearIdStore } from '../../../hooks/useId';
import { overrideAjax } from '../../../tests/ajax';
import attachmentSettings from '../../../tests/ajax/static/context/attachment_settings.json';
import { requireContext } from '../../../tests/helpers';
import { mount } from '../../../tests/reactUtils';
import { f } from '../../../utils/functools';
import { LoadingContext } from '../../Core/Contexts';
import type { SpecifyResource } from '../../DataModel/legacyTypes';
import { deserializeResource } from '../../DataModel/serializers';
import type { Attachment } from '../../DataModel/types';
import { overrideAttachmentSettings } from '../attachments';
import * as Attachments from '../attachments';
import { UploadAttachment } from '../Plugin';
import { testAttachment } from './utils';

requireContext();

async function uploadFileMock(): Promise<SpecifyResource<Attachment>> {
  return deserializeResource(testAttachment);
}

beforeEach(() => {
  clearIdStore();
});

describe('UploadAttachment', () => {
  const testToken = 'testToken';
  const testAttachmentLocation = 'testLocation';

  overrideAjax(
    `/attachment_gw/get_upload_params/`,
    [{ token: testToken, attachmentLocation: testAttachmentLocation }],
    { method: 'POST' }
  );

  test('simple render', async () => {
    jest.spyOn(Attachments, 'uploadFile').mockImplementation(uploadFileMock);
    jest.spyOn(console, 'warn').mockImplementation();
    const handleUploaded = jest.fn();

    overrideAttachmentSettings(attachmentSettings);
    const { asFragment, container, user } = mount(
      <LoadingContext.Provider value={f.void}>
        <UploadAttachment onUploaded={handleUploaded} />
      </LoadingContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();

    const input = Array.from(container.getElementsByTagName('input'))[0];
    const testFile = new File(['Some Text Contents'], 'testName', {
      type: 'text/plain',
    });

    await user.upload(input, testFile);
    fireEvent.change(input, { target: { files: [testFile] } });

    await waitFor(() => {
      expect(handleUploaded).toHaveBeenCalled();
    });
  });
});
