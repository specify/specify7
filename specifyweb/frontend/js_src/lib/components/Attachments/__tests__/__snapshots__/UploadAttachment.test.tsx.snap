// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadAttachment simple render 1`] = `
<DocumentFragment>
  <label
    class="contents"
  >
    <input
      class="sr-only"
      required=""
      type="file"
    />
    <span
      class="
          align-center flex justify-center text-center normal-case
          button hover:brightness-90 dark:hover:brightness-125 bg-[color:var(--secondary-button-color)] text-gray-800 shadow-sm
dark:text-gray-100
          button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm
          h-44 w-full
          
          
        "
    >
      <span>
        Choose a file or drag it here
      </span>
    </span>
  </label>
</DocumentFragment>
`;
